File: backend/app/schemas/analytics.py
"""
Analytics response schemas.
"""

from __future__ import annotations
from typing import Optional
from pydantic import BaseModel


class OverviewResponse(BaseModel):
    total_complaints: int
    open_complaints: int
    resolved_complaints: int
    critical_complaints: int
    total_incidents: int
    open_incidents: int
    resolved_incidents: int
    resolution_rate: float          # 0.0 – 1.0
    average_resolution_time: float  # hours


class CategoryStat(BaseModel):
    category: str
    count: int


class DistrictStat(BaseModel):
    district: str
    count: int


class TrendPoint(BaseModel):
    date: str    # ISO date string "YYYY-MM-DD"
    complaints: int
    incidents: int


class TrendsResponse(BaseModel):
    range: str
    data: list[TrendPoint]


class GeospatialComplaint(BaseModel):
    id: str
    latitude: float
    longitude: float
    category: str
    severity: str
    status: str


class GeospatialIncident(BaseModel):
    id: str
    latitude: float
    longitude: float
    category: str
    severity: str
    status: str
    complaint_count: int


class GeospatialResponse(BaseModel):
    complaints: list[GeospatialComplaint]
    incidents: list[GeospatialIncident]

File: frontend/src/types/analytics.ts
import type { Category, Severity, ComplaintStatus, IncidentStatus } from './common'

export interface AnalyticsOverview {
  total_complaints: number
  open_complaints: number
  resolved_complaints: number
  critical_complaints: number
  total_incidents: number
  open_incidents: number
  resolved_incidents: number
  resolution_rate: number
  average_resolution_time: number
}

export interface CategoryCount {
  category: string
  count: number
}

export interface CategoryDistribution {
  categories: CategoryCount[]
}

export interface DistrictCount {
  district: string
  count: number
}

export interface DistrictDistribution {
  districts: DistrictCount[]
}

export interface TrendPoint {
  date: string
  count: number
}

export interface TrendsResponse {
  range: string
  complaints: TrendPoint[]
  incidents: TrendPoint[]
}

export interface GeospatialComplaint {
  complaint_id: string
  latitude: number
  longitude: number
  category: Category
  severity: Severity
  status: ComplaintStatus
}

export interface GeospatialIncident {
  incident_id: string
  title: string
  latitude: number
  longitude: number
  category: Category
  severity: Severity
  status: IncidentStatus
  complaint_count: number
  district: string
  state: string
  created_at: string
}

export interface GeospatialResponse {
  complaints: GeospatialComplaint[]
  incidents: GeospatialIncident[]
}
